third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -293,37 +405,57 @@ impl InterpreterImpl {
                         )?;
                         continue;
                     }
-                    self.set_new_call_frame::<RTTCheck>(
+
+                    self.set_new_call_frame::<RTTCheck, RTCaches>(
                         &mut current_frame,
                         gas_meter,
                         loader,
                         function,
+                        frame_cache,
                     )?;
                 },
                 ExitCode::CallGeneric(idx) => {
-                    let ty_args = resolver
-                        .instantiate_generic_function(
-                            Some(gas_meter),
-                            idx,
-                            current_frame.function.ty_args(),
-                        )
-                        .map_err(|e| set_err_info!(current_frame, e))?;
-                    let function = resolver
-                        .build_loaded_function_from_instantiation_and_ty_args(idx, ty_args)
-                        .map_err(|e| self.set_location(e))?;
+                    let (function, frame_cache) = if RTCaches::caches_enabled() {
+                        let current_frame_cache = &mut *current_frame.frame_cache.borrow_mut();
 
-                    if RTTCheck::should_perform_checks() {
-                        self.check_friend_or_private_call(&current_frame.function, &function)?;
-                    }
+                        match current_frame_cache.generic_sub_frame_cache.entry(idx) {
+                            btree_map::Entry::Occupied(entry) => {
+                                let entry = entry.get();
+                                (Rc::clone(&entry.0), Rc::clone(&entry.1))
+                            },
+                            btree_map::Entry::Vacant(entry) => {
+                                let function =
+                                    Rc::<LoadedFunction>::new(self.load_generic_function(
+                                        &resolver,
+                                        &current_frame,
+                                        gas_meter,
+                                        idx,
+                                    )?);
+                                let frame_cache = FrameTypeCache::make_rc_for_function(&function);
+
+                                entry.insert((Rc::clone(&function), Rc::clone(&frame_cache)));
```

Comment:
nit: `entry.insert` returns a reference to the new entry already? https://doc.rust-lang.org/std/collections/btree_map/struct.VacantEntry.html#method.insert so you can insert, then clone or map with clones